compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Env.hs

```diff
@@ -58,14 +62,21 @@ getWorld = view world
 
 -- | Type class constraint capturing the needed monadic effects for the
 -- functions manipulating the type checker environment.
-type MonadGamma m = (MonadError Error m, MonadReader Gamma m, MonadState [Warning] m)
+type MonadGamma m = MonadGammaF Gamma m
+type MonadGammaF gamma m = ReaderT gamma (StateT [Warning] (Either Error)) ~ m
```

Comment:
`MonadGamma m` is a constraint over some sort of monad `m`, and is used like

```
myTypechecker :: MonadGamma m => Thing -> m ()
```

If I understand correctly, the other formulation takes some monad m (possibly IO, possible Maybe, possibly Identity) and then wraps it in the layers that would make it a canonical MonadGamma. We'd need to change every type signature to look more like:

```
myTypechecker :: Monad m => Thing -> MonadGamma m ()
```

or further, if we dropped the monad transformer:

```
myTypechecker :: Thing -> MonadGamma ()
```

I decided against that simply because it's a *lot* of changes everywhere, and doesn't necessarily make anything more clear.

Reply:
I find the apparently-polymorphic `m` a bit confusing, instead I'd suggest:
* keep `MonadGamma` as before
* define type synonyms for the transformer stack, not for the constraints
   ```haskell
   type TcMF gamma = ReaderT gamma (StateT [Warning] (Either Error))
   type TcM = TcMF Gamma
   ```
* replace `m` with `TcMF gamma` in the types of `throwWithContextF`, `warnWithContextF` and `withContextF`
* in `Upgrade.hs`, define another synonym for the upgrades transformer stack,
   ```haskell
   type TcUpgradeM = TcMF (Upgrading Gamma)
   ```
* in `Upgrade.hs` type signatures, replace `MonadUpgrade m => m` with `TcUpgradeM` and `MonadGamma m => m` with `TcM`

This avoids changes in unrelated files while simplifying the signatures of functions like `runGammaUnderUpgrades`:
```haskell
runGammaUnderUpgrades :: (MonadGamma m1, MonadUpgrade m2) => Upgrading (m1 a) -> m2 (Upgrading a)
==>
runGammaUnderUpgrades :: Upgrading (TcM a) -> TcUpgradeM (Upgrading a)
```

Reply:
this would also make `TypeFamilies` unnecessary in the other files changed by this PR

Reply:
I've pushed a commit with this change